diff --git a/issue-e/jackson-databind/src/main/java/tools/jackson/databind/deser/jdk/NumberDeserializers.java b/issue-e/jackson-databind/src/main/java/tools/jackson/databind/deser/jdk/NumberDeserializers.java
index ce2303f..63a11c8 100644
--- a/issue-e/jackson-databind/src/main/java/tools/jackson/databind/deser/jdk/NumberDeserializers.java
+++ b/issue-e/jackson-databind/src/main/java/tools/jackson/databind/deser/jdk/NumberDeserializers.java
@@ -476,7 +476,8 @@ public class NumberDeserializers
                         "value outside valid Character range (0x0000 - 0xFFFF)");
             case JsonTokenId.ID_NULL:
                 if (_primitive) {
-                    _verifyNullForPrimitive(ctxt);
+                    Object res = _verifyNullForPrimitive(ctxt);
+                    return (res == null) ? _nullValue : (Character) res;
                 }
                 return (Character) getNullValue(ctxt);
             case JsonTokenId.ID_START_ARRAY:
diff --git a/issue-e/jackson-databind/src/main/java/tools/jackson/databind/deser/jdk/PrimitiveArrayDeserializers.java b/issue-e/jackson-databind/src/main/java/tools/jackson/databind/deser/jdk/PrimitiveArrayDeserializers.java
index 7603bb6..cd10476 100644
--- a/issue-e/jackson-databind/src/main/java/tools/jackson/databind/deser/jdk/PrimitiveArrayDeserializers.java
+++ b/issue-e/jackson-databind/src/main/java/tools/jackson/databind/deser/jdk/PrimitiveArrayDeserializers.java
@@ -282,8 +282,8 @@ public abstract class PrimitiveArrayDeserializers<T>
                             _nuller.getNullValue(ctxt);
                             continue;
                         }
-                        _verifyNullForPrimitive(ctxt);
-                        str = "\0";
+                        Object res = _verifyNullForPrimitive(ctxt);
+                        str = (res == null) ? "\0" : String.valueOf(((Character) res).charValue());
                     } else {
                         CharSequence cs = (CharSequence) ctxt.handleUnexpectedToken(getValueType(ctxt), p);
                         str = cs.toString();
@@ -382,8 +382,8 @@ public abstract class PrimitiveArrayDeserializers<T>
                             _nuller.getNullValue(ctxt);
                             continue;
                         }
-                        _verifyNullForPrimitive(ctxt);
-                        value = false;
+                        Object res = _verifyNullForPrimitive(ctxt);
+                        value = (res == null) ? false : ((Boolean) res).booleanValue();
                     } else {
                         value = _parseBooleanPrimitive(p, ctxt);
                     }
@@ -497,8 +497,8 @@ public abstract class PrimitiveArrayDeserializers<T>
                                 _nuller.getNullValue(ctxt);
                                 continue;
                             }
-                            _verifyNullForPrimitive(ctxt);
-                            value = (byte) 0;
+                            Object res = _verifyNullForPrimitive(ctxt);
+                            value = (res == null) ? (byte) 0 : ((Number) res).byteValue();
                         } else {
                             value = _parseBytePrimitive(p, ctxt);
                         }
@@ -589,8 +589,8 @@ public abstract class PrimitiveArrayDeserializers<T>
                             _nuller.getNullValue(ctxt);
                             continue;
                         }
-                        _verifyNullForPrimitive(ctxt);
-                        value = (short) 0;
+                        Object res = _verifyNullForPrimitive(ctxt);
+                        value = (res == null) ? (short) 0 : ((Number) res).shortValue();
                     } else {
                         value = _parseShortPrimitive(p, ctxt);
                     }
@@ -666,8 +666,8 @@ public abstract class PrimitiveArrayDeserializers<T>
                             _nuller.getNullValue(ctxt);
                             continue;
                         }
-                        _verifyNullForPrimitive(ctxt);
-                        value = 0;
+                        Object res = _verifyNullForPrimitive(ctxt);
+                        value = (res == null) ? 0 : ((Number) res).intValue();
                     } else {
                         value = _parseIntPrimitive(p, ctxt);
                     }
@@ -743,8 +743,8 @@ public abstract class PrimitiveArrayDeserializers<T>
                             _nuller.getNullValue(ctxt);
                             continue;
                         }
-                        _verifyNullForPrimitive(ctxt);
-                        value = 0L;
+                        Object res = _verifyNullForPrimitive(ctxt);
+                        value = (res == null) ? 0L : ((Number) res).longValue();
                     } else {
                         value = _parseLongPrimitive(p, ctxt);
                     }
diff --git a/issue-e/jackson-databind/src/main/java/tools/jackson/databind/deser/std/StdDeserializer.java b/issue-e/jackson-databind/src/main/java/tools/jackson/databind/deser/std/StdDeserializer.java
index 7369e06..f904d40 100644
--- a/issue-e/jackson-databind/src/main/java/tools/jackson/databind/deser/std/StdDeserializer.java
+++ b/issue-e/jackson-databind/src/main/java/tools/jackson/databind/deser/std/StdDeserializer.java
@@ -659,8 +659,8 @@ public abstract class StdDeserializer<T>
                 LogicalType.Integer, Short.TYPE);
         if (act == CoercionAction.AsNull) {
             // 03-May-2021, tatu: Might not be allowed (should we do "empty" check?)
-            _verifyNullForPrimitive(ctxt);
-            return (short) 0;
+            Object res = _verifyNullForPrimitive(ctxt);
+            return (res == null) ? (short) 0 : ((Number) res).shortValue();
         }
         if (act == CoercionAction.AsEmpty) {
             return (short) 0;
@@ -707,8 +707,10 @@ public abstract class StdDeserializer<T>
             // Here regular (strict) accessor is fine
             return p.getIntValue();
         case JsonTokenId.ID_NULL:
-            _verifyNullForPrimitive(ctxt);
-            return 0;
+            {
+                Object res = _verifyNullForPrimitive(ctxt);
+                return (res == null) ? 0 : ((Number) res).intValue();
+            }
         // 29-Jun-2020, tatu: New! "Scalar from Object" (mostly for XML)
         case JsonTokenId.ID_START_OBJECT:
             text = ctxt.extractScalarFromObject(p, this, Integer.TYPE);
@@ -735,8 +737,8 @@ public abstract class StdDeserializer<T>
                 LogicalType.Integer, Integer.TYPE);
         if (act == CoercionAction.AsNull) {
             // 03-May-2021, tatu: Might not be allowed (should we do "empty" check?)
-            _verifyNullForPrimitive(ctxt);
-            return 0;
+            Object res = _verifyNullForPrimitive(ctxt);
+            return (res == null) ? 0 : ((Number) res).intValue();
         }
         if (act == CoercionAction.AsEmpty) {
             return 0;
@@ -863,8 +865,10 @@ public abstract class StdDeserializer<T>
         case JsonTokenId.ID_NUMBER_INT:
             return p.getLongValue();
         case JsonTokenId.ID_NULL:
-            _verifyNullForPrimitive(ctxt);
-            return 0L;
+            {
+                Object res = _verifyNullForPrimitive(ctxt);
+                return (res == null) ? 0L : ((Number) res).longValue();
+            }
         // 29-Jun-2020, tatu: New! "Scalar from Object" (mostly for XML)
         case JsonTokenId.ID_START_OBJECT:
             text = ctxt.extractScalarFromObject(p, this, Long.TYPE);
@@ -891,8 +895,8 @@ public abstract class StdDeserializer<T>
                 LogicalType.Integer, Long.TYPE);
         if (act == CoercionAction.AsNull) {
             // 03-May-2021, tatu: Might not be allowed (should we do "empty" check?)
-            _verifyNullForPrimitive(ctxt);
-            return 0L;
+            Object res = _verifyNullForPrimitive(ctxt);
+            return (res == null) ? 0L : ((Number) res).longValue();
         }
         if (act == CoercionAction.AsEmpty) {
             return 0L;
@@ -1003,8 +1007,10 @@ public abstract class StdDeserializer<T>
         case JsonTokenId.ID_NUMBER_FLOAT:
             return p.getFloatValue();
         case JsonTokenId.ID_NULL:
-            _verifyNullForPrimitive(ctxt);
-            return 0f;
+            {
+                Object res = _verifyNullForPrimitive(ctxt);
+                return (res == null) ? 0f : ((Number) res).floatValue();
+            }
         // 29-Jun-2020, tatu: New! "Scalar from Object" (mostly for XML)
         case JsonTokenId.ID_START_OBJECT:
             text = ctxt.extractScalarFromObject(p, this, Float.TYPE);
@@ -1041,8 +1047,8 @@ public abstract class StdDeserializer<T>
                 LogicalType.Integer, Float.TYPE);
         if (act == CoercionAction.AsNull) {
             // 03-May-2021, tatu: Might not be allowed (should we do "empty" check?)
-            _verifyNullForPrimitive(ctxt);
-            return  0.0f;
+            Object res = _verifyNullForPrimitive(ctxt);
+            return (res == null) ? 0f : ((Number) res).floatValue();
         }
         if (act == CoercionAction.AsEmpty) {
             return  0.0f;
@@ -1123,8 +1129,10 @@ public abstract class StdDeserializer<T>
         case JsonTokenId.ID_NUMBER_FLOAT:
             return p.getDoubleValue();
         case JsonTokenId.ID_NULL:
-            _verifyNullForPrimitive(ctxt);
-            return 0.0;
+            {
+                Object res = _verifyNullForPrimitive(ctxt);
+                return (res == null) ? 0.0 : ((Number) res).doubleValue();
+            }
         // 29-Jun-2020, tatu: New! "Scalar from Object" (mostly for XML)
         case JsonTokenId.ID_START_OBJECT:
             text = ctxt.extractScalarFromObject(p, this, Double.TYPE);
@@ -1161,8 +1169,8 @@ public abstract class StdDeserializer<T>
                 LogicalType.Integer, Double.TYPE);
         if (act == CoercionAction.AsNull) {
             // 03-May-2021, tatu: Might not be allowed (should we do "empty" check?)
-            _verifyNullForPrimitive(ctxt);
-            return  0.0;
+            Object res = _verifyNullForPrimitive(ctxt);
+            return (res == null) ? 0.0 : ((Number) res).doubleValue();
         }
         if (act == CoercionAction.AsEmpty) {
             return  0.0;
